sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -560,9 +565,24 @@ class Analyzer(
     def extract(
         expressions: Seq[NamedExpression]): (Seq[NamedExpression], Seq[NamedExpression]) = {
       val (windowExpressions, regularExpressions) = expressions.partition(hasWindowFunction)
-      // Also need to extract all UnresolvedAttribute from windowExpressions.
+
+      // 1. Substitute expression in windowExpressions with alias in regularExpressions if possible
+      // For example, select sum(key) as s1, sum(sum(key)) over (...) from src group by value,
+      // we need substitute the inner sum(key) with s1.
+      val regularExpressionWithAlias = regularExpressions.collect {
+        case a @ Alias(child, name) => (takeOffAlias(child), a)
+      }.toMap
+
+      val newWindowExpressions = windowExpressions.map {
+        _.transform {
+          case expr if regularExpressionWithAlias.contains(expr) =>
+            regularExpressionWithAlias.get(expr).get.toAttribute
+        }.asInstanceOf[NamedExpression]
+      }
```

Comment:
1 correct me if i misunderstanding, 

>a better way is to extract all expressions from an WindowExpression and push them to the underlying Projects and Aggregates

Do you mean push down all expressions of WindowExpression including self?
      val allexpressions = windowExpressions.flatMap(_.collect {
        case e: Expression => e
      })

then allexpressions here including some expression we do not want to push down,such as
HiveWindowFunction
WindowSpecDefinition
CAST ...

2 i think we can not push down all expressions, see this example:
`select sum(key) as s1, sum(sum(key)) over (...) from src group by value`
we can not push down `key` to underlying aggregate because key is not in group by expressions, this will lead to analysis error.

3 IMO, we should only consider Alias in regularExpressions, user may also use them in windowexpression such as my examples in description above, so we should substitute them.